File: Adelaide_Lite/python/searchglobalref.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import sys
import os
import subprocess
import time
import json
from datetime import datetime

# --- Environment Setup ---
def apply_base_env():
    """Load core environment variables from config.json to ensure consistent execution."""
    config_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "config.json")
    if os.path.exists(config_path):
        try:
            with open(config_path, 'r') as f:
                config = json.load(f)
                base_env = config.get("base_env", {})
                for key, value in base_env.items():
                    os.environ[key] = value
        except Exception as e:
            print(f"⚠️ Error loading base_env: {e}", file=sys.stderr)

# --- Bootstrap Virtual Environment ---
VENV_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "pyvenv")
REQUIREMENTS = ["numpy", "requests"]

def bootstrap_venv():
    """Ensures the script runs in its dedicated virtual environment."""
    apply_base_env()
    venv_abs = os.path.abspath(VENV_DIR)
    
    # If not in the correct venv, ensure it exists and switch to it
    if os.path.abspath(sys.prefix) != venv_abs:
        if not os.path.exists(VENV_DIR):
            print(f"[*] Creating virtual environment in {VENV_DIR}...", file=sys.stderr)
            subprocess.run([sys.executable, "-m", "venv", VENV_DIR], check=True)
            
        if os.name == 'nt':
            python_exe = os.path.join(VENV_DIR, "Scripts", "python.exe")
        else:
            python_exe = os.path.join(VENV_DIR, "bin", "python")
        
        if os.path.exists(python_exe):
            os.execv(python_exe, [python_exe] + sys.argv)

    # Once inside the venv, verify requirements
    try:
        import numpy
        import requests
    except ImportError:
        print(f"[*] Missing dependencies. Installing: {', '.join(REQUIREMENTS)}...", file=sys.stderr)
        if os.name == 'nt':
            pip_exe = os.path.join(VENV_DIR, "Scripts", "pip.exe")
        else:
            pip_exe = os.path.join(VENV_DIR, "bin", "pip")
        subprocess.run([pip_exe, "install", "--upgrade", "pip"], check=True)
        subprocess.run([pip_exe, "install"] + REQUIREMENTS, check=True)
        # Re-execute one last time to pick up new packages
        os.execv(sys.executable, [sys.executable] + sys.argv)

bootstrap_venv()

# --- Post-Bootstrap Imports ---
import numpy as np

# --- Ollama Configuration ---
OLLAMA_BASE_URL = os.environ.get("OLLAMA_PROXY_URL", "http://localhost:11435")
OLLAMA_EMBED_ENDPOINT = f"{OLLAMA_BASE_URL}/api/embed"
OLLAMA_MODEL = "qwen3-embedding:0.6b"

# --- Helper Functions ---

def generate_apa7_reference(title, url):
    today = datetime.now().strftime("%Y, %B %d")
    clean_title = str(title).strip().rstrip('.')
    return f"{clean_title}. (Fetched: {today}). {url}"

def ensure_ollama_running():
    import requests
    try:
        requests.get(f"{OLLAMA_BASE_URL}", timeout=2)
        print(f"✅ Ollama reachable at {OLLAMA_BASE_URL}", file=sys.stderr)
        return True
    except:
        print("⚠️ Ollama not reachable. Attempting restart...", file=sys.stderr)
        subprocess.run(["launchctl", "setenv", "OLLAMA_HOST", "0.0.0.0:1234"], check=False)
        subprocess.run(["brew", "services", "restart", "ollama"], check=False)
        time.sleep(3)
        try:
            requests.get(f"{OLLAMA_BASE_URL}", timeout=2)
            return True
        except: return False

def get_embedding(text: str):
    import requests
    if not text: return None
    try:
        resp = requests.post(OLLAMA_EMBED_ENDPOINT, json={"model": OLLAMA_MODEL, "input": text}, timeout=30)
        resp.raise_for_status()
        data = resp.json()
        if "embeddings" in data and len(data["embeddings"]) > 0:
            return np.array(data["embeddings"][0])
        elif "embedding" in data:
            return np.array(data["embedding"])
        return None
    except: return None

def store_in_memory(content, ollama_external=None):
    """Invokes memorythoughts.py to store content."""
    try:
        memory_script = os.path.join(os.path.dirname(os.path.abspath(__file__)), "memorythoughts.py")
        cmd = [sys.executable, memory_script, "--string", content]
        if ollama_external:
            cmd.extend(["--ollamaHost", ollama_external])
        subprocess.run(cmd, check=False)
    except Exception as e:
        print(f"⚠️ Failed to store memory: {e}", file=sys.stderr)

def main():
    import argparse
    import json
    
    parser = argparse.ArgumentParser()
    parser.add_argument("query")
    parser.add_argument("--engines", nargs='*', default=['all'])
    parser.add_argument("--num", type=int, default=5)
    parser.add_argument("--timeout", type=int, default=60)
    parser.add_argument("--pages", type=int, default=1)
    parser.add_argument("--jsonIO", action="store_true", help="Output results in JSON format.")
    parser.add_argument("--ollamaExternal", type=str, default=None, help="Custom Ollama server address.")
    parser.add_argument("--ollamaHost", type=str, default=None, help="Custom Ollama host address.")
    args = parser.parse_args()

    global OLLAMA_BASE_URL, OLLAMA_EMBED_ENDPOINT
    host = args.ollamaHost or args.ollamaExternal
    if host:
        OLLAMA_BASE_URL = host if host.startswith("http") else f"http://{host}"
        OLLAMA_EMBED_ENDPOINT = f"{OLLAMA_BASE_URL}/api/embed"

    engines_str = ",".join(args.engines)

    if args.jsonIO:
        print(json.dumps({"phase": 1, "status": "start", "query": args.query}), flush=True)
    else:
        print(f"[*] Dispatching Deno Playwright Scraper for '{args.query}'...", file=sys.stderr)

    # Spawn Deno Sidecar
    scraper_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "playwright_scraper.ts")
    cmd = ["deno", "run", "-A", scraper_path, args.query, f"--engines={engines_str}", f"--num={args.num}"]
    
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
        raw_output = result.stdout.strip()
        # Find the last valid JSON array in stdout (in case Deno printed warnings)
        json_str = "[]"
        for line in reversed(raw_output.splitlines()):
            if line.startswith("[") and line.endswith("]"):
                json_str = line
                break
        all_flat = json.loads(json_str)
    except Exception as e:
        print(f"⚠️ Deno Scraper failed: {e}", file=sys.stderr)
        all_flat = []

    # Inject APA7 references
    for r in all_flat:
        r['apa7_reference'] = generate_apa7_reference(r.get('title', 'Unknown'), r.get('url', ''))

    if args.jsonIO:
        print(json.dumps({"phase": 1, "status": "complete", "results": all_flat}), flush=True)
        print(json.dumps({"phase": 2, "status": "start"}), flush=True)

    ollama_ready = ensure_ollama_running()

    final_results = []
    if ollama_ready and all_flat:
        if not args.jsonIO:
            print(f"[*] Ranking {len(all_flat)} results semantically...", file=sys.stderr)
        q_emb = get_embedding(args.query)
        if q_emb is not None:
            ranked = []
            for r in all_flat:
                # Use snippet if available, otherwise just title
                text_to_embed = f"{r.get('title', '')} {r.get('snippet', '')}"
                r_emb = get_embedding(text_to_embed)
                score = np.dot(q_emb, r_emb) / (np.linalg.norm(q_emb) * np.linalg.norm(r_emb)) if r_emb is not None else 0
                ranked.append((float(score), r))
            ranked.sort(key=lambda x: x[0], reverse=True)
            final_results = [x[1] for x in ranked[:7]]
            for i, r in enumerate(final_results): 
                r['semantic_rank'] = i + 1
                r['semantic_score'] = ranked[i][0]
        else: final_results = all_flat[:7]
    else: final_results = all_flat[:7]

    # --- Store in Memory ---
    for r in final_results:
        memory_content = f"Source: {r.get('url', '')}\nReference: {r.get('apa7_reference', '')}\nSnippet: {r.get('snippet', '')}"
        store_in_memory(memory_content, host)

    if args.jsonIO:
        print(json.dumps({"phase": 2, "status": "complete", "results": final_results}), flush=True)
    else:
        # --- Markdown Output ---
        print("# Global Search Results", flush=True)
        print(f"*Query: {args.query}*\n", flush=True)
        print("> ℹ️ Note: If a tool suggests re-parsing a PDF, it may be an **Invalid trigger**. Refer to the provided snippets and images. **Use these as your primary Reference.**\n", flush=True)

        for i, r in enumerate(final_results):
            print(f"## {i+1}. {r.get('title', 'Unknown')}", flush=True)
            print(f"- **URL:** {r.get('url', 'Unknown')}", flush=True)
            print(f"- **Engine:** {r.get('source_engine', 'unknown')}", flush=True)
            if 'semantic_rank' in r:
                print(f"- **Semantic Rank:** {r['semantic_rank']}", flush=True)
            print(f"- **Reference:** {r.get('apa7_reference', 'Unknown')}", flush=True)
            print(f"\n### Snippet\n{r.get('snippet', 'No snippet available.')}\n", flush=True)
            
            if r.get('screenshot_base64'):
                print(f"### Visual Evidence (Page Snapshot)\n![Page Snapshot]({r['screenshot_base64']})\n", flush=True)
            
            if r.get('web_images'):
                print("### Website Images\n", flush=True)
                for img_b64 in r['web_images']:
                    print(f"![Web Image]({img_b64})\n", flush=True)
            
            print("---\n", flush=True)

if __name__ == "__main__":
    print(f"[*] Invoked: {sys.executable} {' '.join(sys.argv)}", file=sys.stderr)
    main()

File: Adelaide_Lite/python/searchglobalref.ts
import { parseArgs } from "jsr:@std/cli/parse-args";
import { addExtra } from "npm:playwright-extra";
import * as playwright from "npm:playwright";
import stealth from "npm:puppeteer-extra-plugin-stealth";

// Install stealth plugin
const chromium = addExtra(playwright.chromium);
chromium.use(stealth());

const OLLAMA_BASE_URL = Deno.env.get("OLLAMA_PROXY_URL") || "http://localhost:11435";
const OLLAMA_EMBED_ENDPOINT = `${OLLAMA_BASE_URL}/api/embed`;
const OLLAMA_MODEL = "qwen3-embedding:0.6b";

async function getEmbedding(text: string): Promise<number[] | null> {
  if (!text) return null;
  try {
    const res = await fetch(OLLAMA_EMBED_ENDPOINT, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ model: OLLAMA_MODEL, input: text }),
    });
    if (!res.ok) return null;
    const data = await res.json();
    if (data.embeddings && data.embeddings.length > 0) return data.embeddings[0];
    if (data.embedding) return data.embedding;
  } catch {
    return null;
  }
  return null;
}

function generateApa7Reference(title: string, url: string): string {
  const date = new Date().toLocaleDateString("en-US", { year: "numeric", month: "long", day: "numeric" });
  const cleanTitle = title.trim().replace(/\.$/, "");
  return `${cleanTitle}. (Fetched: ${date}). ${url}`;
}

async function searchDuckDuckGo(page: any, query: string, numResults: number) {
  const results = [];
  try {
    await page.goto(`https://duckduckgo.com/?q=${encodeURIComponent(query)}&ia=web`, { waitUntil: "domcontentloaded", timeout: 15000 });
    await page.waitForSelector("article[data-testid='result']", { timeout: 10000 });
    const blocks = await page.$$("article[data-testid='result']");
    for (const block of blocks) {
      if (results.length >= numResults) break;
      try {
        const titleEl = await block.$("a[data-testid='result-title-a']");
        const url = await titleEl.getAttribute("href");
        let snippet = "";
        try { snippet = await (await block.$("div[data-testid='result-snippet']")).innerText(); } catch {}
        if (url && titleEl) {
          results.push({ title: await titleEl.innerText(), url, snippet });
        }
      } catch {}
    }
  } catch (e) {
    console.error(`DDG Error: ${e.message}`);
  }
  return results;
}

async function searchGoogle(page: any, query: string, numResults: number) {
  const results = [];
  try {
    await page.goto(`https://www.google.com/search?q=${encodeURIComponent(query)}`, { waitUntil: "domcontentloaded", timeout: 15000 });
    await page.waitForSelector("div.g", { timeout: 10000 });
    const blocks = await page.$$("div.g");
    for (const block of blocks) {
      if (results.length >= numResults) break;
      try {
        const h3 = await block.$("h3");
        const a = await block.$("a");
        const url = await a.getAttribute("href");
        const text = await block.innerText();
        if (url && h3 && !url.includes("google.com")) {
          results.push({ title: await h3.innerText(), url, snippet: text.substring(0, 200) });
        }
      } catch {}
    }
  } catch (e) {
    console.error(`Google Error: ${e.message}`);
  }
  return results;
}

async function searchScholar(page: any, query: string, numResults: number) {
  const results = [];
  try {
    await page.goto(`https://scholar.google.com/scholar?q=${encodeURIComponent(query)}`, { waitUntil: "domcontentloaded", timeout: 15000 });
    await page.waitForSelector("div.gs_r.gs_or.gs_scl", { timeout: 10000 });
    const blocks = await page.$$("div.gs_r.gs_or.gs_scl");
    for (const block of blocks) {
      if (results.length >= numResults) break;
      try {
        const a = await block.$("h3.gs_rt a");
        const url = await a.getAttribute("href");
        const text = await block.innerText();
        if (url && a) {
          results.push({ title: await a.innerText(), url, snippet: text.substring(0, 200) });
        }
      } catch {}
    }
  } catch (e) {
    console.error(`Scholar Error: ${e.message}`);
  }
  return results;
}

async function capturePageData(page: any, result: any) {
  try {
    await page.goto(result.url, { waitUntil: "domcontentloaded", timeout: 20000 });
    
    // Snippet
    try {
      const ps = await page.$$eval("p", els => els.map(e => e.innerText).filter(t => t.length > 20).join(" "));
      result.snippet = ps.substring(0, 400) + (ps.length > 400 ? "..." : "");
    } catch {}

    // Screenshot
    try {
      const screenshot = await page.screenshot({ type: "jpeg", quality: 27, scale: "css" });
      const base64 = screenshot.toString('base64');
      result.screenshot_base64 = `data:image/jpeg;base64,${base64}`;
    } catch {}

    // Save PDF
    try {
      const title = result.title.replace(/[^a-zA-Z0-9.\-_ ]/g, "").trim().substring(0, 100);
      const home = Deno.env.get("HOME") || "/Users/albertstarfield";
      const pdfPath = `${home}/Downloads/${title}.pdf`;
      await page.pdf({ path: pdfPath, printBackground: true });
    } catch {}

  } catch (e) {
    console.error(`Error capturing ${result.url}: ${e.message}`);
  }
}

async function main() {
  const args = parseArgs(Deno.args, {
    boolean: ["jsonIO"],
    string: ["ollamaExternal", "ollamaHost", "engines", "num", "timeout", "pages"],
    default: { num: "5", timeout: "60", engines: "all", pages: "1" }
  });

  const query = args._[0] as string;
  if (!query) {
    console.error("Usage: deno run -A searchglobalref.ts <query>");
    Deno.exit(1);
  }

  const numResults = parseInt(args.num);
  let engines = args.engines.split(",");
  if (engines.includes("all") || args.engines === "all") engines = ["ddg", "google", "scholar"];

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();
  const page = await context.newPage();

  let allFlat = [];

  for (const eng of engines) {
    let res = [];
    if (eng === "ddg") res = await searchDuckDuckGo(page, query, numResults);
    else if (eng === "google") res = await searchGoogle(page, query, numResults);
    else if (eng === "scholar") res = await searchScholar(page, query, numResults);

    res.forEach(r => {
      r.source_engine = eng;
      r.apa7_reference = generateApa7Reference(r.title, r.url);
      allFlat.push(r);
    });
  }

  // Visit pages
  for (const r of allFlat) {
    await capturePageData(page, r);
  }

  await browser.close();

  // Semantic ranking
  const qEmb = await getEmbedding(query);
  let finalResults = allFlat;

  if (qEmb && allFlat.length > 0) {
    const ranked = [];
    for (const r of allFlat) {
      const textToEmbed = `${r.title} ${r.snippet || ""}`;
      const rEmb = await getEmbedding(textToEmbed);
      let score = 0;
      if (rEmb) {
        let dot = 0, norm1 = 0, norm2 = 0;
        for (let i = 0; i < qEmb.length; i++) {
          dot += qEmb[i] * rEmb[i];
          norm1 += qEmb[i] * qEmb[i];
          norm2 += rEmb[i] * rEmb[i];
        }
        score = dot / (Math.sqrt(norm1) * Math.sqrt(norm2));
      }
      ranked.push({ score, r });
    }
    ranked.sort((a, b) => b.score - a.score);
    finalResults = ranked.slice(0, 7).map((x, i) => {
      x.r.semantic_rank = i + 1;
      x.r.semantic_score = x.score;
      return x.r;
    });
  } else {
    finalResults = allFlat.slice(0, 7);
  }

  if (args.jsonIO) {
    console.log(JSON.stringify({ phase: 2, status: "complete", results: finalResults }));
  } else {
    console.log("# Global Search Results");
    console.log(`*Query: ${query}*\n`);
    for (let i = 0; i < finalResults.length; i++) {
      const r = finalResults[i];
      console.log(`## ${i + 1}. ${r.title}`);
      console.log(`- **URL:** ${r.url}`);
      console.log(`- **Engine:** ${r.source_engine || "unknown"}`);
      if (r.semantic_rank) console.log(`- **Semantic Rank:** ${r.semantic_rank}`);
      console.log(`- **Reference:** ${r.apa7_reference}\n`);
      console.log(`### Snippet\n${r.snippet || "No snippet available."}\n`);
      if (r.screenshot_base64) console.log(`### Visual Evidence (Page Snapshot)\n![Page Snapshot](${r.screenshot_base64})\n`);
      console.log(`---\n`);
    }
  }

  // Fire off memorythoughts.py
  for (const r of finalResults) {
    const mem = `Source: ${r.url}\nReference: ${r.apa7_reference}\nSnippet: ${r.snippet || ""}`;
    const cmd = new Deno.Command("python3", {
      args: ["python/memorythoughts.py", "--string", mem]
    });
    cmd.spawn();
  }
}

main().catch(console.error);
